docs: setup for writing guides (#138)
sapphire-project/framework: .github/CONTRIBUTING.md

When writing guides there are some rules to follow:

1. All files should have the `.mdx` file extension.
2. Folder names are allowed to have spaces.
3. Each word in a folder name should started with a Capital letter.
4. File names should _never_ have spaces.
5. File names that consist of multiple words should be PascalCased.
6. Information in a guide page should be generally useful to the majority of people.
    - The single exception to this goes to files in the "Advanced" folder, which can cover any type of advanced topics.
7. There are several variables that will be replaced when parsing the `.mdx` file for the documentation website. Variables are denoted by the pattern `{@variableName (parameter)}`. The supported variables are:
    - `{@branch}` - for the current branch name.

>     - `{@link <TSDoc symbol name>}` - wherein `TSDoc symbol name` is a reference to some TypeScript interface / class / function etc parsed from TSDoc. This will link the API page of that symbol.

Reviewer: ```suggestion
    - `{@link <TSDoc symbol name>}` - wherein `<TSDoc symbol name>` is a reference to some TypeScript interface / class / function etc parsed from TSDoc. This will link the API page of that symbol.
```

    - `{@guide <Guide file name>}` - wherein `<Guide file name>` is the name of another guide file in this same project _without file extension_. For example linking to `CreatingArguments.mdx` would be `{@guide CreatingArguments}`.
8. For variables that are often reused between pages, please add them to `constants.mdx` in the `guides` folder. You can reference these variables just like how you would import code with ES Modules: `import { something } from '../constants.mdx`. **Be sure to include the `.mdx` file extension!**

    - The `constants.mdx` file can also have functions to manipulate some parameters prior to being injected into the `.mdx` file. It should be noted however that in order to use the data returned by a function in a property of a JSX tag (i.e. in the `href` property of an `<a>` tag) that it should be called at the top of the `.mdx` file and stored to its own, local variable. For example:

>     ```md

Reviewer 2: Might want to add a wrong example.

    import { doSomethingCool } from '../constants.mdx';

    export const parsedCool = doSomethingCool('veryCool');

    <a href={parsedCool}>cool</a>
    ```

Lastly when writing guides we highly recommend that you use the [VSCode] editor and install the following extensions. This will allow you to get both syntax highlighting as well as live preview.

> 1. [MDX]

Reviewer 2: Possibility to add a `extensions.json` in the `.vscode` folder to show the "recommendations" for this repo;
![image](https://user-images.githubusercontent.com/5418114/105421108-d6b8d880-5c41-11eb-96c7-6a3a8d7ef5e3.png)

Reviewer 2: More things could be added to benefit from this (take a look at your `devDependencies` and see if there are more helpful plugins.)

# Contributing

## Workflow

> 1. Fork, clone, and select the **main** branch.
> 2. Create a new branch in your fork.

Reviewer 3: ```suggestion
1. Fork and clone this repository.
2. Create a new branch in your fork based off the **main** branch.
```

3. Make your changes.
4. Commit your changes, and push them.
5. Submit a Pull Request [here]!

## Contributing to the guides

> The guides for all Sapphire Project use [mdx] formatting. At its core this is the exact same as regular markdown, and it supports all your regular GitHub flavoured markdown syntax. Where it differs from this regular markdown however, is that it _also_ supports inline JSX tags. These are used for advanced and optimal layout on `<docs site address pending>`. If you want to learn how and what kind of JSX you can embed into your [mdx] then please consult the [mdx documentation][]

Reviewer 3: ```suggestion
The guides for all Sapphire Project use [mdx] formatting. At its core this is the exact same as regular markdown, and it supports all your regular GitHub flavoured markdown syntax. Where it differs from this regular markdown however, is that it _also_ supports inline JSX tags. These are used for advanced and optimal layout on `<docs site address pending>`. If you want to learn how and what kind of JSX you can embed into your [mdx] then please consult the [mdx documentation]
```

When writing guides there are some rules to follow:

1. All files should have the `.mdx` file extension.

> 2. Folder names are allowed to have spaces.

Reviewer 4: Would it not be prevalent to, in line with filesystem conventions and ease of use, use lowercased folders with hyphens instead of spaces and then have the site convert that when reading the guides pages? I really do not think it's worth violating filesystem conventions for this, especially not when large companies do it so much already and get a fair bit of hate from dev communities for it.

3. Each word in a folder name should started with a Capital letter.
4. File names should _never_ have spaces.
5. File names that consist of multiple words should be PascalCased.
6. Information in a guide page should be generally useful to the majority of people.
    - The single exception to this goes to files in the "Advanced" folder, which can cover any type of advanced topics.
7. There are several variables that will be replaced when parsing the `.mdx` file for the documentation website. Variables are denoted by the pattern `{@variableName (parameter)}`. The supported variables are:

>     - `{@branch}` - for the current branch name.
>     - `{@link <TSDoc symbol name>}` - wherein `<TSDoc symbol name>` is a reference to some TypeScript interface / class / function etc parsed from TSDoc. This will link the API page of that symbol.
>     - `{@typedoc <TSDoc interface/type/class name>}` - wherein `<TSDoc interface/type/class name>` is the name of a TypeScript interface, class or type. This will inline all the properties of that interface/type/class as a table, with links back to their API documentation page.
>     - `{@guide <Guide file name>}` - wherein `<Guide file name>` is the name of another guide file in this same project _without file extension_. For example linking to `CreatingArguments.mdx` would be `{@guide CreatingArguments}`.

Reviewer 4: ```suggestion
    - `{@branch}` - the current branch name.
    - `{@link <TSDoc symbol name>}` - links to the TSDoc generated API page for `<TSDoc symbol name>`, which should be a reference to some TypeScript interface, class, function, et cetera, parsed from TSDoc.
    - `{@typedoc <TSDoc interface/type/class>}` - this does the same as the above _for interfaces, types and classes exclusively_, except it also inlines all the properties of the TSDoc object in a markdown table that links back to their respective API documentation pages.
    - `{@guide <Guide file name>}` - links to another guide in the project, where `<Guide file name>` is the filename _without the file extension_ of another guide file in the same scope. For example, to link to a file named `CreatingArguments.mdx`, you would use `{@guide CreatingArguments}`.
```

> 8. For variables that are often reused between pages, please add them to `constants.mdx` in the `guides` folder. You can reference these variables just like how you would import code with ES Modules: `import { something } from '../constants.mdx`. **Be sure to include the `.mdx` file extension!**
> 
>     - The `constants.mdx` file can also have functions to manipulate some parameters prior to being injected into the `.mdx` file. It should be noted however that in order to use the data returned by a function in a property of a JSX tag (i.e. in the `href` property of an `<a>` tag) that it should be called at the top of the `.mdx` file and stored to its own, local variable. For example:

Reviewer 4: ```suggestion
8. For variables that are often reused between pages, please add them to `constants.mdx` in the `guides` folder. You can reference these variables just like how you would import an ECMAScript Module: `import { something } from '../constants.mdx`. **Be sure to include the `.mdx` file extension!**
    - The constants file can also store functions that manipulate parameters before being injected into a guide file. However, to use a function's return values in a JSX tag's properties (exempli gratia, in the reference/`href` property of an anchor/`<a>` tag), you should call it at the top of the guide file and store the result in a local variable. For example:
```

    ```md
    <!-- !!!DO THIS!!! -->

    import { doSomethingCool } from '../constants.mdx';

    export const parsedCool = doSomethingCool('veryCool');

    <a href={parsedCool}>cool</a>

    <!-- !!!DO NOT DO THIS!!! -->

    import { doSomethingCool } from '../constants.mdx';

    <a href={doSomethingCool('veryCool')}>cool</a>
    ```

> Lastly when writing guides we highly recommend that you use the [VSCode] editor and install the following extensions. This will allow you to get both syntax highlighting as well as live preview.

Reviewer 4: ```suggestion
Lastly, when writing guides we highly recommend that you use [Visual Studio Code][vscode] and install the following extensions. These will give you both syntax highlighting and live previews for your MDX. However, your editor of choice likely has similar extensions. Failing that, though, there are tools to assist you both online and in CLI, such as the [ESLint plugin for MDX][eslint-mdx].
```

> 1. [MDX]

Reviewer 3: iirc it works this way
```suggestion
1. [MDX][mdx-vscode]
```

> 2. [MDX Preview][]

Reviewer 3: ```suggestion
2. [MDX Preview]
```

-   Everything in @sapphire/framework should be generally useful for the majority of users. Don't let that stop you if you've got a good concept though, as your idea still might be a great addition.
-   Everything should be shard compliant. If code you put in a pull request would break when sharding, break other things from supporting sharding, or is incompatible with sharding; then you will need to think of a way to make it work with sharding in mind before the pull request will be accepted and merged.
-   Everything should follow [OOP paradigms] and generally rely on behaviour over state where possible. This generally helps methods be predictable, keeps the codebase simple and understandable, reduces code duplication through abstraction, and leads to efficiency and therefore scalability.
-   Everything should follow our ESLint rules as closely as possible, and should pass lint tests even if you must disable a rule for a single line.
-   Scripts that are to be ran outside of the scope of the bot should be added to [scripts] directory and should be in the `.mjs` file format.

<!-- Link Dump -->

<!-- Guides -->

> [mdx]: https://marketplace.visualstudio.com/items?itemName=silvenon.mdx

Reviewer 3: ```suggestion
[mdx-vscode]: https://marketplace.visualstudio.com/items?itemName=silvenon.mdx
```

> [mdx-vscode]: https://marketplace.visualstudio.com/items?itemName=silvenon.mdx
> [mdx preview]: https://marketplace.visualstudio.com/items?itemName=xyc.vscode-mdx-preview
> [mdx]: https://mdxjs.com
> [mdx documentation]: https://mdxjs.com/getting-started
> [vscode]: https://code.visualstudio.com

Reviewer 4: ```suggestion
[mdx-vscode]: https://marketplace.visualstudio.com/items?itemName=silvenon.mdx
[mdx preview]: https://marketplace.visualstudio.com/items?itemName=xyc.vscode-mdx-preview
[mdx]: https://mdxjs.com
[mdx documentation]: https://mdxjs.com/getting-started
[vscode]: https://code.visualstudio.com
[eslint-mdx]: https://github.com/mdx-js/eslint-mdx
```
